contracts/src/v0.8/ccip/pools/SiloedLockReleaseTokenPool.sol

```diff
@@ -0,0 +1,294 @@
+    bool isSiloed; // ──────╯ Whether funds should be isolated from all other chains or shared amongst all non-siloed chains.
+  }
+
+  /// @notice The configuration for each chain that is siloed, or not. By default chains are not siloed.
+  mapping(uint64 remoteChainSelector => SiloConfig) internal s_chainConfigs;
+
+  constructor(
+    IERC20 token,
+    uint8 localTokenDecimals,
+    address[] memory allowlist,
+    address rmnProxy,
+    address router
+  ) TokenPool(token, localTokenDecimals, allowlist, rmnProxy, router) {}
+
+  /// @notice Locks the token in the pool
+  /// @dev The _validateLockOrBurn check is an essential security check
+  function lockOrBurn(
+    Pool.LockOrBurnInV1 calldata lockOrBurnIn
+  ) external virtual override returns (Pool.LockOrBurnOutV1 memory) {
+    _validateLockOrBurn(lockOrBurnIn);
+
+    // If funds need to be siloed, update internal accounting;
+    if (s_chainConfigs[lockOrBurnIn.remoteChainSelector].isSiloed) {
+      s_chainConfigs[lockOrBurnIn.remoteChainSelector].tokenBalance += lockOrBurnIn.amount;
+    }
+    // If the messages is going to a chain without siloed funds, update state accounting accordingly.
+    else {
+      s_unsiloedTokenBalance += lockOrBurnIn.amount;
+    }
+
+    emit Locked(msg.sender, lockOrBurnIn.amount);
+
+    return Pool.LockOrBurnOutV1({
+      destTokenAddress: getRemoteToken(lockOrBurnIn.remoteChainSelector),
+      destPoolData: _encodeLocalDecimals()
+    });
+  }
+
+  /// @notice Release tokens from the pool to the recipient
+  /// @dev The _validateReleaseOrMint check is an essential security check
+  /// @dev If the releaseOrMintIn amount is greater than available liquidity, the function will revert as a security
+  /// measure to prevent funds from a Silo being released by another chain.
+  function releaseOrMint(
+    Pool.ReleaseOrMintInV1 calldata releaseOrMintIn
+  ) external virtual override returns (Pool.ReleaseOrMintOutV1 memory) {
+    _validateReleaseOrMint(releaseOrMintIn);
+
+    // Calculate the local amount
+    uint256 localAmount =
+      _calculateLocalAmount(releaseOrMintIn.amount, _parseRemoteDecimals(releaseOrMintIn.sourcePoolData));
+
+    // Save gas by using storage instead of memory as a value may need to be updated.
+    SiloConfig storage remoteConfig = s_chainConfigs[releaseOrMintIn.remoteChainSelector];
+
+    // Prevent A silent underflow by explicitly ensuring that enough funds are available to release
+    uint256 availableLiquidity = remoteConfig.isSiloed ? remoteConfig.tokenBalance : s_unsiloedTokenBalance;
+    if (localAmount > availableLiquidity) revert InsufficientLiquidity(availableLiquidity, localAmount);
+
+    // Tracking balances independently by chain is a security measure to prevent liquidity for one chain from being
+    // released by another chain. Since all tokens are stored locally, and not isolated by contract,
+    // it must be ensured that the remoteChainSelector can be trusted.
```

Comment:
not sure how we can trust a chainselector? You mean "don't mess up the config"? I think that kinda holds for all contacts :p

Reply:
It's more like we're making the assumption that a message has not been tampered with in the functioning of CCIP up until this point, but if that no longer holds true well we've got bigger problems. Seems like a superfluous comment now. Will remove.